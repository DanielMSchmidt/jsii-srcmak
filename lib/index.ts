export * from './srcmak';
export * from './compile';
export * from './options';