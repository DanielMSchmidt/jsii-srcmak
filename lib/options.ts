export interface CommonOptions {
  /**
   * The relative path of the .ts entrypoint within the source directory.
   * @default "index.ts"
   */
  entrypoint?: string;

  /**
   * List of directories that include node modules to symlink into the compiled
   * package. For example, if your generated code references some library, you
   * should include it's module directory in here.
   */
  moduleDirs?: string[];

  /**
   * Path to output the .jsii file output.
   * @default - jsii file is not emitted.
   */
  outputJsii?: string;
}

export interface PythonOptions extends CommonOptions {
  /**
   * The name of the the python module to generate. If omitted python will not be generated.
   */
  pythonName?: string;
}

export type Options = PythonOptions; /* | JavaOptions | DotNetOptions */