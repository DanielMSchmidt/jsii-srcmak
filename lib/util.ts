import * as fs from 'fs-extra';
import * as path from 'path';
import * as os from 'os';
import { spawn, SpawnOptions } from 'child_process';

export async function withTempDir(dirname: string, closure: (dir: string) => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s.'));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure(workdir);

    if (!process.env.WITH_TEMP_DIR_RETAIN) {
      await fs.remove(parent);
    } else {
      console.error(`retained temp dir: ${parent}`);
    }
  } catch(e) {
    console.error(`retained temp dir due to an error: ${parent}`);
    throw e;
  } finally {
    process.chdir(prevdir);
  }
}

export async function exec(moduleName: string, args: string[] = [], options: SpawnOptions = { }) {
  return new Promise((ok, fail) => {

    const opts: SpawnOptions = {
      ...options,
      stdio: [ 'inherit', 'pipe', 'pipe' ],
    };
    const child = spawn(process.execPath, [ moduleName, ...args ], opts);

    const data = new Array<Buffer>();
    child.stdout?.on('data', chunk => data.push(chunk));
    child.stderr?.on('data', chunk => data.push(chunk));

    const newError = (message: string) => new Error([
      message,
      `COMMAND: ${moduleName} ${args.join(' ')}`,
      `WORKDIR: ${path.resolve(options.cwd ?? '.')}`,
      '------------------------------------------------------------------------------------',
      Buffer.concat(data).toString('utf-8'),
      '------------------------------------------------------------------------------------',
    ].join('\n'));

    child.once('error', err => {
      throw newError(`jsii compilation failed. error: ${err.message}`);
    });

    child.once('exit', code => {
      if (code === 0) {
        return ok();
      }
      else {
        return fail(newError(`jsii compilation failed with non-zero exit code: ${code}`));
      }
    });
  });
}
