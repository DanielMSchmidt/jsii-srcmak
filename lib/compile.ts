import * as fs from 'fs-extra';
import * as path from 'path';
import { exec } from './util';
import { Options } from './options';

const compilerModule = require.resolve('jsii/bin/jsii');

/**
 * Compiles the source files in `workdir` with jsii.
 */
export async function compile(workdir: string, options: Options) {
  const args = [ '--silence-warnings', 'reserved-word' ];
  const entrypoint = options.entrypoint ?? 'index.ts';

  if (path.extname(entrypoint) !== '.ts') {
    throw new Error(`jsii entrypoint must be a .ts file: ${entrypoint}`);
  }

  // path to entrypoint without extension
  const basepath = path.join(path.dirname(entrypoint), path.basename(entrypoint, '.ts'));

  // jsii modules to include
  const modules = options.modules ?? [];

  const targets: Record<string, any> = { };

  const deps: Record<string, string> = { };
  for (const mod of modules) {
    if (mod.startsWith('@types/')) {
      continue;
    }

    deps[mod] = '*';
  }

  const pkg = {
    name: 'generated',
    version: '0.0.0',
    author: 'generated@generated.com',
    main: `${basepath}.js`,
    types: `${basepath}.d.ts`,
    license: 'Apache-2.0',
    repository: { url: 'http://generated', type: 'git' },
    jsii: {
      outdir: 'dist',
      targets: targets,
    },
    dependencies: deps,
    peerDependencies: deps,
  };

  if (options.pythonName) {
    targets.python = {
      distName: 'generated',
      module: options.pythonName,
    };
  }

  for (const mod of modules) {
    const sourcedir = path.dirname(require.resolve(`${mod}/package.json`));
    await fs.mkdirp(path.join(workdir, path.join('node_modules', path.dirname(mod))));
    await fs.ensureSymlink(sourcedir, path.join(workdir, 'node_modules', mod));
  }

  await fs.writeFile(path.join(workdir, 'package.json'), JSON.stringify(pkg, undefined, 2));

  await exec(compilerModule, args, {
    cwd: workdir,
  });
}
