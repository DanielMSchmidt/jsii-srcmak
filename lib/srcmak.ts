import * as fs from 'fs-extra';
import * as path from 'path';
import { withTempDir, exec } from './util';
import { compile } from './compile';
import { Options } from './options';

const pacmakModule = require.resolve('jsii-pacmak/bin/jsii-pacmak');

export async function srcmak(srcdir: string, targetdir: string, options: Options) {
  srcdir = path.resolve(srcdir);
  targetdir = path.resolve(targetdir);

  await withTempDir('jsii-codemak', async () => {
    // copy sources to temp directory
    await fs.copy(srcdir, '.');

    // perform jsii compilation
    await compile('.', options);

    // run pacmak to generate code
    await exec(pacmakModule, [ '--code-only' ]);

    // extract code based on selected languages
    if (options.pythonName) {
      const reldir = options.pythonName.replace(/\./g, '/'); // replace "." with "/"
      const source = path.resolve(`dist/python/src/${reldir}`);
      const target = path.join(targetdir, reldir);
      await fs.move(source, target, { overwrite: true });
    }
  });
}
