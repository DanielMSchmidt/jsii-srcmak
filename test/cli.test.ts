import { execFileSync } from 'child_process';
import * as path from 'path';
import * as fs from 'fs-extra';
import { mkdtemp } from '../lib/util';
import { snapshotDirectory } from './util';

const program = require.resolve('../bin/jsii-srcmak');
const srcdir = path.join(__dirname, '..', 'example');

test('no arguments - fails with help', () => {
  expect(() => srcmakcli()).toThrow(/Invalid number of arguments. expecting a single positional argument./);
});

test('invalid entrypoint', () => {
  expect(() => srcmakcli(srcdir)).toThrow(/unable to find typescript entrypoint/);
});

test('compile only', () => {
  srcmakcli(
    srcdir,
    '--entrypoint lib/main.ts',
  );
});

test('jsii output', () => {
  const path = '/tmp/my-jsii-output.json';

  srcmakcli(srcdir,
    '--entrypoint lib/main.ts',
    '--jsii-path ${path}',
  );

  expect(JSON.parse(fs.readFileSync(path, 'utf-8'))).toMatchSnapshot();
});

test('fails if only one python option is given', () => {
  expect(() => srcmakcli(srcdir,
    '--entrypoint lib/main.ts',
    '--python-module-name module',
  )).toThrow(/--python-outdir is required if --python-module-name is specified/);

  expect(() => srcmakcli(srcdir,
    '--entrypoint lib/main.ts',
    '--python-outdir dir',
  )).toThrow(/--python-module-name is required if --python-outdir is specified/);
});

test('python output', async () => {
  await mkdtemp(async outdir => {
    srcmakcli(srcdir,
      '--entrypoint lib/main.ts',
      `--python-outdir ${outdir}`,
      '--python-module-name my.python.module',
    );

    expect(await snapshotDirectory(outdir, [ 'generated@0.0.0.jsii.tgz' ])).toMatchSnapshot();
  });
});

test('dependencies', async () => {
  await mkdtemp(async srcdir => {
    await fs.writeFile(path.join(srcdir, 'index.ts'), `
      import * as fs from 'fs';
      fs.writeFileSync('foo.bar', 'hello');
    `);

    // resolve against *this* executable
    const dep = path.dirname(require.resolve('@types/node/package.json'));
    srcmakcli(srcdir, `--dep ${dep}`);
  });
});

function srcmakcli(...args: string[]) {
  const a = args.join(' ');
  execFileSync(program, a ? a.split(' ') : []);
}