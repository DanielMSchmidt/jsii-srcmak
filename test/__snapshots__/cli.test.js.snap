// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsii output 1`] = `
Object {
  "author": Object {
    "name": "generated@generated.com",
    "roles": Array [
      "author",
    ],
  },
  "description": "generated",
  "fingerprint": "YzbxW50rbVtKyO3xyXOGChedDt8pdGdCRYTETM1PW7k=",
  "homepage": "http://generated",
  "jsiiVersion": "1.5.0 (build 46538f8)",
  "license": "Apache-2.0",
  "name": "generated",
  "repository": Object {
    "type": "git",
    "url": "http://generated",
  },
  "schema": "jsii/0.10.0",
  "targets": Object {
    "js": Object {
      "npm": "generated",
    },
  },
  "types": Object {
    "generated.Calculator": Object {
      "assembly": "generated",
      "docs": Object {
        "summary": "A sophisticaed multi-language calculator.",
      },
      "fqn": "generated.Calculator",
      "initializer": Object {},
      "kind": "class",
      "locationInModule": Object {
        "filename": "lib/main.ts",
        "line": 19,
      },
      "methods": Array [
        Object {
          "docs": Object {
            "summary": "Adds the two operands.",
          },
          "locationInModule": Object {
            "filename": "lib/main.ts",
            "line": 24,
          },
          "name": "add",
          "parameters": Array [
            Object {
              "docs": Object {
                "summary": "operands.",
              },
              "name": "ops",
              "type": Object {
                "fqn": "generated.Operands",
              },
            },
          ],
          "returns": Object {
            "type": Object {
              "primitive": "number",
            },
          },
        },
        Object {
          "docs": Object {
            "summary": "Multiplies the two operands.",
          },
          "locationInModule": Object {
            "filename": "lib/main.ts",
            "line": 40,
          },
          "name": "mul",
          "parameters": Array [
            Object {
              "docs": Object {
                "summary": "operands.",
              },
              "name": "ops",
              "type": Object {
                "fqn": "generated.Operands",
              },
            },
          ],
          "returns": Object {
            "type": Object {
              "primitive": "number",
            },
          },
        },
        Object {
          "docs": Object {
            "summary": "Subtracts the two operands.",
          },
          "locationInModule": Object {
            "filename": "lib/main.ts",
            "line": 32,
          },
          "name": "sub",
          "parameters": Array [
            Object {
              "docs": Object {
                "summary": "operands.",
              },
              "name": "ops",
              "type": Object {
                "fqn": "generated.Operands",
              },
            },
          ],
          "returns": Object {
            "type": Object {
              "primitive": "number",
            },
          },
        },
      ],
      "name": "Calculator",
    },
    "generated.Operands": Object {
      "assembly": "generated",
      "datatype": true,
      "docs": Object {
        "summary": "Math operands.",
      },
      "fqn": "generated.Operands",
      "kind": "interface",
      "locationInModule": Object {
        "filename": "lib/main.ts",
        "line": 4,
      },
      "name": "Operands",
      "properties": Array [
        Object {
          "abstract": true,
          "docs": Object {
            "summary": "Left-hand side operand.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "lib/main.ts",
            "line": 8,
          },
          "name": "lhs",
          "type": Object {
            "primitive": "number",
          },
        },
        Object {
          "abstract": true,
          "docs": Object {
            "summary": "Right-hand side operand.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "lib/main.ts",
            "line": 13,
          },
          "name": "rhs",
          "type": Object {
            "primitive": "number",
          },
        },
      ],
    },
  },
  "version": "0.0.0",
}
`;

exports[`python output 1`] = `
Object {
  "my/python/module/__init__.py": "import abc
import builtins
import datetime
import enum
import typing

import jsii
import jsii.compat
import publication

from ._jsii import *


class Calculator(metaclass=jsii.JSIIMeta, jsii_type=\\"generated.Calculator\\"):
    \\"\\"\\"A sophisticaed multi-language calculator.\\"\\"\\"
    def __init__(self) -> None:
        jsii.create(Calculator, self, [])

    @jsii.member(jsii_name=\\"add\\")
    def add(self, *, lhs: jsii.Number, rhs: jsii.Number) -> jsii.Number:
        \\"\\"\\"Adds the two operands.

        :param lhs: Left-hand side operand.
        :param rhs: Right-hand side operand.
        \\"\\"\\"
        ops = Operands(lhs=lhs, rhs=rhs)

        return jsii.invoke(self, \\"add\\", [ops])

    @jsii.member(jsii_name=\\"mul\\")
    def mul(self, *, lhs: jsii.Number, rhs: jsii.Number) -> jsii.Number:
        \\"\\"\\"Multiplies the two operands.

        :param lhs: Left-hand side operand.
        :param rhs: Right-hand side operand.
        \\"\\"\\"
        ops = Operands(lhs=lhs, rhs=rhs)

        return jsii.invoke(self, \\"mul\\", [ops])

    @jsii.member(jsii_name=\\"sub\\")
    def sub(self, *, lhs: jsii.Number, rhs: jsii.Number) -> jsii.Number:
        \\"\\"\\"Subtracts the two operands.

        :param lhs: Left-hand side operand.
        :param rhs: Right-hand side operand.
        \\"\\"\\"
        ops = Operands(lhs=lhs, rhs=rhs)

        return jsii.invoke(self, \\"sub\\", [ops])


@jsii.data_type(jsii_type=\\"generated.Operands\\", jsii_struct_bases=[], name_mapping={'lhs': 'lhs', 'rhs': 'rhs'})
class Operands():
    def __init__(self, *, lhs: jsii.Number, rhs: jsii.Number) -> None:
        \\"\\"\\"Math operands.

        :param lhs: Left-hand side operand.
        :param rhs: Right-hand side operand.
        \\"\\"\\"
        self._values = {
            'lhs': lhs,
            'rhs': rhs,
        }

    @builtins.property
    def lhs(self) -> jsii.Number:
        \\"\\"\\"Left-hand side operand.\\"\\"\\"
        return self._values.get('lhs')

    @builtins.property
    def rhs(self) -> jsii.Number:
        \\"\\"\\"Right-hand side operand.\\"\\"\\"
        return self._values.get('rhs')

    def __eq__(self, rhs) -> bool:
        return isinstance(rhs, self.__class__) and rhs._values == self._values

    def __ne__(self, rhs) -> bool:
        return not (rhs == self)

    def __repr__(self) -> str:
        return 'Operands(%s)' % ', '.join(k + '=' + repr(v) for k, v in self._values.items())


__all__ = [
    \\"Calculator\\",
    \\"Operands\\",
]

publication.publish()
",
  "my/python/module/_jsii/__init__.py": "import abc
import builtins
import datetime
import enum
import typing

import jsii
import jsii.compat
import publication

__jsii_assembly__ = jsii.JSIIAssembly.load(\\"generated\\", \\"0.0.0\\", __name__[0:-6], \\"generated@0.0.0.jsii.tgz\\")

__all__ = [
    \\"__jsii_assembly__\\",
]

publication.publish()
",
  "my/python/module/py.typed": "
",
}
`;
